src/ray/core_worker/pubsub/pubsub_coordinator.cc

```diff
@@ -0,0 +1,199 @@
+}
+
+void Subscriber::QueueMessage(const ObjectID &object_id, bool try_publish) {
+  mailbox_.push_back(object_id);
+  if (try_publish) {
+    PublishIfPossible();
+  }
+}
+
+bool Subscriber::PublishIfPossible(bool force) {
+  if (long_polling_reply_callback_ == nullptr) {
+    return false;
+  }
+  if (force || mailbox_.size() > 0) {
+    long_polling_reply_callback_(mailbox_);
+    long_polling_reply_callback_ = nullptr;
+    mailbox_.clear();
+    return true;
+  }
+  return false;
+}
+
+void PubsubCoordinator::Connect(const NodeID &subscriber_node_id,
+                                LongPollConnectCallback long_poll_connect_callback) {
+  absl::MutexLock lock(&mutex_);
+  RAY_LOG(DEBUG) << "Long polling connection is initiated by " << subscriber_node_id;
+  RAY_CHECK(long_poll_connect_callback != nullptr);
+
+  if (is_node_dead_(subscriber_node_id)) {
+    std::vector<ObjectID> empty_ids;
+    // Reply to the request so that it is properly GC'ed.
+    long_poll_connect_callback(empty_ids);
+    return;
+  }
+
+  if (subscribers_.count(subscriber_node_id) == 0) {
+    subscribers_.emplace(subscriber_node_id, std::make_shared<Subscriber>());
+  }
+  auto subscriber = subscribers_[subscriber_node_id];
+
+  // Since the long polling connection is synchronous between the client and coordinator,
+  // when it connects, the connection shouldn't have existed.
+  RAY_CHECK(subscriber->Connect(std::move(long_poll_connect_callback)));
+  subscriber->PublishIfPossible();
+}
+
+void PubsubCoordinator::RegisterSubscription(const NodeID &subscriber_node_id,
+                                             const ObjectID &object_id) {
+  absl::MutexLock lock(&mutex_);
+  RAY_LOG(DEBUG) << "object id " << object_id << " is subscribed by "
+                 << subscriber_node_id;
+  if (is_node_dead_(subscriber_node_id)) {
+    return;
+  }
+
+  if (subscribers_.count(subscriber_node_id) == 0) {
+    subscribers_.emplace(subscriber_node_id, std::make_shared<Subscriber>());
+  }
+  subscription_index_.AddEntry(object_id, subscriber_node_id);
+}
+
+void PubsubCoordinator::Publish(const ObjectID &object_id) {
+  absl::MutexLock lock(&mutex_);
+  for (const auto &subscriber_id :
+       subscription_index_.GetSubscriberIdsByObjectId(object_id)) {
+    auto it = subscribers_.find(subscriber_id);
+    RAY_CHECK(it != subscribers_.end());
+    auto subscriber = it->second;
+    subscriber->QueueMessage(object_id);
+  }
+}
+
+int PubsubCoordinator::UnregisterSubscriber(const NodeID &subscriber_node_id) {
+  absl::MutexLock lock(&mutex_);
+  int erased = subscription_index_.EraseSubscriber(subscriber_node_id);
+  // Publish messages before removing the entry. Otherwise, it can have memory leak.
+  auto it = subscribers_.find(subscriber_node_id);
+  if (it == subscribers_.end()) {
+    return erased;
+  }
+  auto subscriber = it->second;
+  // Remove the long polling connection because otherwise, there's memory leak.
+  subscriber->PublishIfPossible(/*force=*/true);
```

Comment:
Supernit: Ditto
```suggestion
  // Remove the long polling connection because otherwise, there's memory leak.
  it->second->PublishIfPossible(/*force=*/true);
```